src/vision/calib/Calibration.cpp

```diff
@@ -0,0 +1,210 @@
+#include "MonoCalibrator.hpp"
+#include "wfov_camera_msgs/WFOVImage.h"
+#include <cv_bridge/cv_bridge.h>
+#include <sensor_msgs/image_encodings.h>
+#include <string>
+#include <vector>
+
+using namespace cv_bridge;
+
+Settings s;
+
+bool calibFinished = false;
+vector<vector<Point2f>> imagePoints;
+Mat cameraMatrix, distCoeffs;
+Size imageSize;
+int mode = CAPTURING;
+clock_t prevTimestamp = 0;
+const Scalar RED(0, 0, 255), GREEN(0, 255, 0);
+const char ESC_KEY = 27;
+
+void imageCallback(const wfov_camera_msgs::WFOVImage::ConstPtr& msg)
+{
+    ROS_DEBUG_STREAM("Callback, Mode: " << mode);
+    Mat view = toCvShare(msg->image, msg,
+                         sensor_msgs::image_encodings::BGR8)->image;
+
+    if (mode == CAPTURING && imagePoints.size() >= (size_t)s.nrFrames)
+    {
+        ROS_DEBUG_STREAM("Calibrating and saving during capture");
+        if (runCalibrationAndSave(s, imageSize, cameraMatrix, distCoeffs,
+                                  imagePoints))
+            mode = CALIBRATED;
+        else
+            mode = DETECTION;
+    }
+
+    if (view.empty())
+    {
+        ROS_DEBUG_STREAM("Empty view");
+        if (mode != CALIBRATED && !imagePoints.empty())
+        {
+            runCalibrationAndSave(s, imageSize, cameraMatrix, distCoeffs,
+                                  imagePoints);
+        }
+        calibFinished = true;
+        ROS_DEBUG_STREAM("Exiting after empty view. Mode: " << mode);
+        ros::shutdown();
+        return;
+    }
+
+    imageSize = view.size();
+    if (s.flipVertical) flip(view, view, 0);
+
+    vector<Point2f> pointBuf;
+
+    bool found;
+
+    int chessBoardFlags = CALIB_CB_ADAPTIVE_THRESH | CALIB_CB_NORMALIZE_IMAGE;
+
+    if (!s.useFisheye)
+    {
+        chessBoardFlags |= CALIB_CB_FAST_CHECK;
+    }
+
+    switch(s.calibrationPattern)
+    {
+        case Settings::CHESSBOARD:
+            ROS_DEBUG_STREAM("Finding Chessboard Corners");
+            found = findChessboardCorners(view, s.boardSize, pointBuf,
+                                          chessBoardFlags);
+            break;
+        case Settings::CIRCLES_GRID:
+            found = findCirclesGrid(view, s.boardSize, pointBuf);
+            break;
+        case Settings::ASYMMETRIC_CIRCLES_GRID:
+            found = findCirclesGrid(view, s.boardSize, pointBuf,
+                                    CALIB_CB_ASYMMETRIC_GRID);
+            break;
+        default:
+            found = false;
+            break;
+    }
+
+    if (found)
+    {
+        ROS_DEBUG_STREAM("Pattern found");
+        if (s.calibrationPattern == Settings::CHESSBOARD)
+        {
+            Mat viewGray;
+            cvtColor(view, viewGray, COLOR_BGR2GRAY);
+            cornerSubPix(viewGray, pointBuf, Size(11, 11), Size(-1, -1),
+                         TermCriteria(TermCriteria::EPS+TermCriteria::COUNT,
+                                      30, 0.1));
+        }
```

Comment:
Can you please document why the chessboard is a special case and what specifically you're doing to it?